### PR TITLE
Integrate Jira

### PR DESCRIPTION
**Pull request by hoxhaeris**

### CODE REVIEW

File: cmd/release-controller/main.go [hunk 1]
```diff
@@ -307,14 +325,14 @@ func (o *options) Run() error {
 	if err != nil {
 		return fmt.Errorf("unable to create image client: %v", err)
 	}
-	klog.V(4).Infof("1: %v", time.Now().Sub(start))
+	klog.V(4).Infof("1: %v", time.Since(start))
```
Comment (bradmwilliams) [inline cmd/release-controller/main.go:328]: All these lines can actually be removed now.

File: cmd/release-controller/main.go [hunk 2]
```diff
@@ -402,7 +405,26 @@ func (o *options) Run() error {
...
 		c.bugzillaErrorMetrics = bugzillaErrorMetrics
 	}
-	klog.V(4).Infof("8: %v", time.Now().Sub(start))
+
+	if o.VerifyJira {
+		ghClient, err := o.github.GitHubClient(false)
+		if err != nil {
+			return fmt.Errorf("Failed to create github client: %v", err)
+		}
+		ghClient.Throttle(o.githubThrottle, 0)
```
Comment (bradmwilliams) [inline cmd/release-controller/main.go:414]: We probably only need to initialize a single github client that can be shared by both Bugzilla and Jira.

File: cmd/release-controller/main.go [hunk 3]
```diff
@@ -174,7 +190,8 @@ func main() {
 	flagset.StringVar(&opt.ListenAddr, "listen", opt.ListenAddr, "The address to serve metrics on")
 
 	flagset.BoolVar(&opt.VerifyBugzilla, "verify-bugzilla", opt.VerifyBugzilla, "Update status of bugs fixed in accepted release to VERIFIED if PR was approved by QE.")
-	flagset.IntVar(&opt.githubThrottle, "github-throttle", 0, "Maximum number of GitHub requests per hour. Used by bugzilla verifier.")
+	flagset.BoolVar(&opt.VerifyBugzilla, "verify-jira", opt.VerifyJira, "Update status of bugs fixed in accepted release to VERIFIED if PR was approved by QE.")
```
[the remark this answers is not in the document]
Comment (hoxhaeris) [reply cmd/release-controller/main.go]: Fixed

File: cmd/release-controller/main.go (hunk 1 above)
Comment (hoxhaeris) [reply cmd/release-controller/main.go:328]: Remove all `klog.V(4).Infof("1: %v", time.Since(start))` ?

File: cmd/release-controller/main.go (hunk 2 above)
Comment (hoxhaeris) [reply cmd/release-controller/main.go]: Fixed. Github client is initialised separately.

File: pkg/jira/jira.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	if err != nil {
+		return unlabeledPRs, fmt.Errorf("unable to get labels for github pull %s/%s#%d: %w", extPR.org, extPR.repo, extPR.prNum, err)
+	}
+	return unlabeledPRs, nil
+}
+
+func (c *Verifier) verifyExtPRs(issue *jiraBaseClient.Issue, extPRs []pr, errs []error, tagName string) (bool, string, []error, bool) {
+	var success bool
+	message := fmt.Sprintf("Bugfix included in accepted release %s", tagName)
```
[the remark this answers is not in the document]
Comment (hoxhaeris) [reply pkg/jira/jira.go]: Fixed

File: pkg/jira/jira.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+			var newErr error
+			unlabeledPRs, newErr = c.ghUnlabeledPRs(extPR)
+			if newErr != nil {
+				errs = append(errs, newErr)
+				issueErrs = append(issueErrs, newErr)
+			}
+		}
+	}
+	if len(unlabeledPRs) > 0 || len(issueErrs) > 0 {
+		message = fmt.Sprintf("%s\nBug will not be automatically moved to VERIFIED for the following reasons:", message)
```
[the remark this answers is not in the document]
Comment (hoxhaeris) [reply pkg/jira/jira.go]: Fixed

File: pkg/jira/jira.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+			}
+		}
+		if success {
+			// TODO - check the status. Bugzilla equivalent is "VERIFIED"
+			klog.V(4).Infof("Updating issue %s (current status %s) to VERIFIED status", issue.ID, issue.Fields.Status.Name)
+			if _, err := c.jiraClient.UpdateIssue(&jiraBaseClient.Issue{Fields: &jiraBaseClient.IssueFields{Status: &jiraBaseClient.Status{Name: jira.StatusVerified}}}); err != nil {
+				errs = append(errs, fmt.Errorf("failed to update status for issue %s: %w", issue.ID, err))
+			}
+		} else {
+			klog.V(4).Infof("Bug %s (current status %s) not approved by QA contact", issue.ID, issue.Fields.Status.Name)
```
[the remark this answers is not in the document]
Comment (hoxhaeris) [reply pkg/jira/jira.go]: Fixed

File: pkg/release-controller/types.go [hunk 7]
```diff
@@ -163,6 +163,8 @@ type ReleasePublish struct {
 	ImageStreamRef *PublishStreamReference `json:"imageStreamRef"`
 	// VerifyBugs marks bugs fixed by this tag as VERIFIED in bugzilla if the QA contact reviewed and approved the bugfix PR
 	VerifyBugs *PublishVerifyBugs `json:"verifyBugs"`
+	// TODO - check if this is correct
+	VerifyIssues *PublishVerifyIssues `json:"verifyIssues"`
```
[the remark this answers is not in the document]
Comment (hoxhaeris) [reply pkg/release-controller/types.go]: Fixed

File: pkg/release-controller/types.go [hunk 8]
```diff
@@ -206,6 +208,24 @@ type VerifyBugsTagInfo struct {
...
 
+// PublishVerifyIssues marks bugs fixed by this tag as VERIFIED in Jira if the QA contact reviewed and approved the bugfix PR
+type PublishVerifyIssues struct {
+	// PreviousRelease points to the last release created before the imagestream
+	// being published was created. It is used to verify bugs on the oldest tag
+	// in the release being published.
+	PreviousReleaseTag *VerifyIssuesTagInfo `json:"previousReleaseTag"`
+}
+
+// VerifyIssuesTagInfo contains the necessary data to get a tag reference as needed in the bugzilla verification support.
```
[the remark this answers is not in the document]
Comment (hoxhaeris) [reply pkg/release-controller/types.go]: Fixed

File: pkg/release-controller/types.go [hunk 9]
```diff
@@ -515,6 +535,11 @@ const (
 	// processed by the BugzillaVerifier
 	ReleaseAnnotationBugsVerified = "release.openshift.io/bugs-verified"
 
+	// ReleaseAnnotationIssuesVerified indicates whether the release has been
+	// processed by the BugzillaVerifier
+	//TODO - check the path and if there will be an TagReference annotation for Issues
+	ReleaseAnnotationIssuesVerified = "release.openshift.io/issues-verified"
+
```
[the remark this answers is not in the document]
Comment (hoxhaeris) [reply pkg/release-controller/types.go]: Fixed

**Review by bradmwilliams — APPROVED** [review]

**Pull request merged**